<a href="https://colab.research.google.com/github/JosenildoJunior/StatPyDataScience/blob/main/Correla%C3%A7%C3%A3o_e_Regress%C3%A3o_Linear.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Correlação**

A correlação é uma medida estatística que descreve a relação entre duas variáveis. Ela avalia a direção e a força da associação entre essas variáveis.

Os dados podem ser representados por pares ordenados (x, y), sendo x a variável independente (ou explanatória) e y a variável dependente (ou resposta).

## **Como funciona ?**

A correlação é calculada utilizando técnicas estatísticas, como o coeficiente de correlação de Pearson, Spearman ou Kendall. Esses coeficientes variam de -1 a 1, onde -1 indica uma correlação negativa perfeita, 0 indica nenhuma correlação e 1 indica uma correlação positiva perfeita.

Força da Correlação:

- 0.00 a 0.19 (ou -0.19): Correlação muito fraca.
- 0.20 a 0.39 (ou -0.39): Correlação fraca.
- 0.40 a 0.59 (ou -0.59): Correlação moderada.
- 0.60 a 0.79 (ou -0.79): Correlação forte.
- 0.80 a 1.00 (ou -1.00): Correlação muito forte.

Vamos observar um exemplo para entendermos melhor o assunto

Considere um conjunto de dados que represente as notas de dois alunos (Aluno A e Aluno B) em duas disciplinas diferentes (Matemática e Ciências). As notas são representadas em uma escala de 0 a 10.

Crie um DataFrame com os seguintes dados:

- Aluno A - Matemática: [8, 7, 6, 9, 8]
- Aluno B - Matemática: [7, 6, 8, 9, 10]
- Aluno A - Ciências: [9, 8, 7, 9, 8]
- Aluno B - Ciências: [8, 9, 7, 8, 9]

Calcule a correlação entre as notas do Aluno A e do Aluno B em ambas as disciplinas.

Imprima o resultado da correlação para cada disciplina.

Interprete os resultados.

In [ ]:
# Manipulação de dados
import pandas as pd

# Funções estatísticas
import scipy.stats as st

# Vamos criar um DataFrame de exemplo
dados = {
    'Aluno_a_matematica': [8, 7, 6, 9, 8],
    'Aluno_b_matematica': [7, 6, 8, 9, 10],
    'Aluno_a_ciencias': [9, 8, 7, 9, 8],
    'Aluno_b_ciencias': [8, 9, 7, 8, 9]
}

# Convertendo para um df
df = pd.DataFrame(dados)

# Calculando a correlação entre as duas variáveis
correlacao_mat, p_valor_mat = st.pearsonr(df['Aluno_a_matematica'], df['Aluno_b_matematica'])

# Calculando a correlação entre as duas variáveis
correlacao_cien, p_valor_cien = st.pearsonr(df['Aluno_a_ciencias'], df['Aluno_b_ciencias'])

# Exibindo o resultado
print(f"A correlação entre as variáveis em matématica é de: {correlacao_mat} já em ciências é de: {correlacao_cien}")

A correlação entre as variáveis em matématica é de: 0.4160251471689218 já em ciências é de: 0.2857142857142857


Observando a correlação nas notas de matemática, percebemos uma correlação moderada positiva, indicando que, à medida que o valor de uma variável aumenta, a outra variável tende a aumentar também.

A correlação das notas de ciências é fraca e positiva. Nesse caso, o mesmo se aplica: à medida que uma variável aumenta, a outra tende a aumentar.

É interessante notar que, nesse caso, também poderíamos utilizar o método 'corr' para calcular a correlação de **Pearson**. Também seria possível utilizarmos a função 'corrcoef' para realizar o mesmo cálculo.

Além do Pearson, também existe o **Spearman**, que é uma medida de associação entre duas variáveis ordinais ou de intervalo. Ele avalia a relação monotônica entre as variáveis, ou seja, se uma variável aumenta, a outra tende a aumentar (ou diminuir), independentemente da magnitude do aumento (ou diminuição).

Por outro lado, o coeficiente de correlação de Pearson avalia especificamente a relação linear entre duas variáveis, o que implica que ele assume uma relação constante entre os aumentos das variáveis. Se a relação entre as variáveis não for linear, o Pearson pode não capturar essa relação adequadamente.

Vamos observar a diferença entre o Spearman e o Pearson aplicando o Spearman nesta mesma questão.

In [ ]:
# Importando o Spearman
from scipy.stats import spearmanr

# Calculando a correlação de Spearman
correlacao_mat_spear, p_valor_mat_spear = spearmanr(df['Aluno_a_matematica'], df['Aluno_b_matematica'])

# Calculando a correlação de Spearman
correlacao_cien_spear, p_valor_cien_spear = spearmanr(df['Aluno_a_ciencias'], df['Aluno_b_ciencias'])

# Exibindo o resultado
print(f"A correlação Spearman entre as variáveis em matématica é de: {correlacao_mat_spear} já em ciências é de: {correlacao_cien_spear}")

A correlação Spearman entre as variáveis em matématica é de: 0.46169025843831935 já em ciências é de: 0.1111111111111111


Aplicamos o Spearman nos mesmos dados e obtivemos resultados diferentes. Isso se deve porque, embora o Spearman também esteja medindo a direção e a força da correlação, ele está fazendo isso para a relação monótona, ou seja, ele avalia se, quando uma variável aumenta, a outra tende a aumentar ou diminuir consistentemente, mas não necessariamente em uma taxa constante.

Logo, chegamos à conclusão de que:

- Correlação de Pearson: Avalia a relação linear entre variáveis contínuas.
- Correlação de Spearman: Avalia a relação monótona entre variáveis, sendo menos sensível a outliers e distribuições não-normais.

### **Qual utilizar Pearson ou Spearman ?**

Para sabermos essa questão, primeiro precisamos entender a diferença entre elas:

- A principal diferença entre a correlação de Pearson e a correlação de Spearman é que a correlação de Pearson avalia a associação linear entre duas variáveis, enquanto a correlação de Spearman avalia a associação monótona, o que a torna mais robusta em relação a outliers e a distribuições não-normais.

Agora que já entendemos a diferença, fica mais fácil saber quando utilizar uma ou outra, concorda?

- Você deve usar a correlação de Pearson quando suas variáveis são contínuas e seguem uma distribuição normal, e você está interessado na relação linear entre elas. Por outro lado, você deve usar a correlação de Spearman quando suas variáveis são ordinais, quando a relação entre elas não é linear ou quando a distribuição dos dados é não normal.

### **Teste de hipóteses**

Basicamente, podemos utilizar o teste de hipóteses para determinarmos se o coeficiente de correlação amostral fornece evidências suficientes de que o coeficiente de correlação populacional é significativo. Dito isso, nossas hipóteses seguem a seguinte fórmula:

- H0: ρ = 0 (Não há correlação significativa)
- Ha: ρ ≠ 0 (Correlação significativa)

Então, digamos que o p-valor seja menor que o nível de significância. Dessa forma, podemos rejeitar a hipótese nula. Portanto, podemos dizer que a correlação é estatisticamente significativa.

Nesse caso, vou aplicar o teste nas correlações de Spearman. Vamos utilizar um nível de significância de 0,05.

In [ ]:
# Observando os p-valores
print(f"p-valor das notas em matemática: {p_valor_mat_spear}\np-valor das notas em ciências: {p_valor_cien_spear}")

p-valor das notas em matemática: 0.43376615616902786
p-valor das notas em ciências: 0.8588205736427662


Em ambos os casos, o p-valor foi maior que nosso nível de significância, logo, falhamos em rejeitar a hipótese nula. Dito isso, com base em nosso teste a um nível de significância de 0,05, falhamos em rejeitar a hipótese nula. Logo, podemos concluir que não existem evidências de que as correlações sejam estatisticamente significativas em nossa população.

# **Regressão Linear**

A regressão é uma técnica que nos permite quantificar e inferir a relação de uma variável dependente (que será a nossa variável resposta) com variáveis independentes (que são as nossas variáveis explicativas).

Pontos importantes a se ter em mente:

- **Variáveis dependentes**: Estas são as variáveis que estamos tentando explicar ou prever usando um modelo. Os valores dessas variáveis dependem de outras e normalmente as chamamos de 'y'.

- **Variáveis independentes**: Estas são as variáveis que usamos para explicar ou prever as variáveis dependentes. Elas também são conhecidas como variáveis preditoras e normalmente as chamamos de 'X'.

- **Coeficiente de determinação \( R^2 \)**: O coeficiente de determinação \( R^2 \) (R-quadrado) é uma métrica fundamental para avaliar o ajuste de um modelo de regressão aos dados. Se o \( R^2 \) de um modelo de regressão for, por exemplo, 0,328, isso significa que o modelo consegue explicar aproximadamente 32,8% da variabilidade dos dados. Em outras palavras, cerca de 32,8% da variação na variável dependente pode ser explicada pelas variáveis independentes incluídas no modelo.

Regressão linear simples x Regressão múltipla:

- **Regressão linear simples**: Envolvendo apenas uma variável independente, essa regressão visa traçar uma linha reta que se ajuste aos dados para modelar a relação da melhor forma possível entre a variável dependente e a única independente.

- **Regressão linear múltipla**: Envolvendo duas ou mais variáveis independentes, essa regressão visa obter um plano que forme o melhor ajuste possível, minimizando a soma dos quadrados das diferenças entre os valores observados e os valores previstos pelo modelo linear múltiplo.

Em **resumo**, a regressão linear simples analisa a relação entre duas variáveis, enquanto a regressão linear múltipla analisa a relação entre uma variável dependente e duas ou mais variáveis independentes.

**Mínimos Quadrados Ordinários (OLS)**

O OLS é um tipo de análise de regressão. Para aplicarmos essa análise, é necessário respeitar alguns pontos importantes, tais como:

- As **variáveis independentes** podem ser contínuas ou categóricas.

- A **variável dependente** deve ser contínua.

**Valores observados** x **Valores ajustados**

- **Valores observados**: São os valores registrados durante nosso estudo ou experimento, juntamente com os valores das variáveis independentes. Esses valores são denotados usando Y.

- **Valores ajustados**: São os valores que nosso modelo prevê. Valores previstos e valores ajustados são sinônimos. É comum denotarmos valores ajustados como Ŷ.

**O que são resíduos e qual a sua importância?**

Utilizamos os resíduos para avaliar o quão bem está indo nosso modelo. Precisamos analisar a diferença entre os valores observados e os valores ajustados, pois essa diferença representa o erro no modelo.

Um resíduo é a distância entre o valor observado e o valor ajustado correspondente. A fórmula de cálculo dos resíduos é:

- **Resíduos** = Valor observado – Valor ajustado

Graficamente falando, os resíduos são as distâncias verticais entre os valores e os valores ajustados.

No entanto, não podemos simplesmente somar os resíduos, pois os valores negativos e positivos se anularão. É aí que entra o **OLS**, pois essa análise elevará os **resíduos ao quadrado**. Dessa forma, poderemos somá-los sem anulá-los. Logo, um ponto de dados com um resíduo de 3 terá um erro quadrado de 9.

Então o **OLS** desenha a linha que minimiza a **soma dos erros ao quadrado (SSE)**.

**O que é a soma dos erros ao quadrado (SSE)?**

Basicamente, o SSE é uma medida de variabilidade. O SSE aumenta conforme os pontos se afastam da linha ajustada. Como os cálculos usam diferenças quadradas, a variância está em unidades quadradas em vez das unidades originais dos dados. Para um dado conjunto, valores menores de SSE sinalizam que as observações estão mais próximas dos valores ajustados.








## **Exemplos**

Até aqui, já vimos bastante conceitos. Vamos observar uma parte um pouco mais prática agora. Dito isso, seguiremos as seguintes etapas:

Importação dos dados

Análise rápida dos dados (Nesse caso, queremos partir logo para a regressão, mas em situações reais seria necessário realizar uma análise mais minuciosa nos dados)

Preparar os dados (Observar os tipos de dados presentes, se existe algum valor ausente)

Dividir os dados (Dividir entre treino e teste)

Treinar o modelo (Treinar o modelo nas variáveis selecionadas)

Avaliar o modelo (Avaliar o desempenho do modelo)

Nosso objetivo é realizar uma análise de regressão linear simples ou múltipla para prever os preços das casas com base em diferentes atributos.

### **Importando os dados**

In [3]:
# Manipulação de dados
import pandas as pd

# URL do conjunto de dados no GitHub
url = 'https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv'

# Carregando o conjunto de dados
df = pd.read_csv(url)

# Exibindo as primeiras linhas
df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

**Dicionário dos dados**

- CRIM: Taxa de criminalidade per capita por cidade.

- ZN: Proporção de terrenos residenciais zoneados para lotes com mais de 25.000 pés quadrados.

- INDUS: Proporção de acres de negócios não comerciais por cidade.

- CHAS: Variável fictícia de Charles River (1 se o trecho limita o rio; 0 caso contrário).

- NOX: Concentração de óxidos nítricos (partes por 10 milhões).

- RM: Número médio de quartos por habitação.

- AGE: Proporção de unidades ocupadas pelo proprietário construídas antes de 1940.

- DIS: Distâncias ponderadas para cinco centros de emprego em Boston.

- RAD: Índice de acessibilidade às rodovias radiais.

- TAX: Taxa de imposto sobre a propriedade de valor total por US$ 10.000.

- PTRATIO: Proporção de alunos por professor na cidade.

- B: 1000(Bk - 0,63)^2, onde Bk é a proporção de pessoas de descendência afro-americana por cidade.

- LSTAT: Percentual da população de status baixo.

- MEDV: Valor mediano das casas ocupadas pelos proprietários em US$ 1000.

Nosso processo de carregamento foi um sucesso, então vamos partir para a próxima parte.

## **Análise rápida dos dados**

Como realizar uma análise completa não é o nosso objetivo atual, vamos observar as correlações mais fortes com a variável preço, pois iremos utilizá-la futuramente.

In [10]:
# Calculando as correlações entre medv
correlations = df.corrwith(df['medv'])

# Exibir as correlações
print(correlations)

medv       1.000000
rm         0.695360
zn         0.360445
b          0.333461
dis        0.249929
chas       0.175260
age       -0.376955
rad       -0.381626
crim      -0.388305
nox       -0.427321
tax       -0.468536
indus     -0.483725
ptratio   -0.507787
lstat     -0.737663
dtype: float64


A variável que apresentou maior correlação foi a medv. Dito isso, vamos seguir adiante.

In [12]:
# Observando as dimenções do nosso df
df.shape

(506, 14)

Temos 506 linhas e 14 colunas. Vamos seguir para a preparação de dados.

## **Preparando os dados**

Vamos observar se existe algum valor ausente nas nossas variaveis

In [11]:
# Observando valores ausentes
df.isnull().sum()

crim       0
zn         0
indus      0
chas       0
nox        0
rm         0
age        0
dis        0
rad        0
tax        0
ptratio    0
b          0
lstat      0
medv       0
dtype: int64

Como podemos observar não existe nenhum valor ausente no nosso df

## **Dividir os dados**

Para conclusão dessa etapa sera necessario a importação de alguns pacotes, dito isso vamos seguir com as importações

In [ ]:
# Divisão dos dados
from sklearn.model_selection import train_test_split

# Regressão linear
from sklearn.linear_model import LinearRegression

# Métricas de validação do modelo
from sklearn.metrics import r2_score, mean_squared_error

Agora vamos realizar a divisão dos nossos dados entre variaveis dependente e independente, como vimos anteriormente a variavel dependente é o valor que queremos prever e a variavel independente são os dados que serão utilizados para prever a variavel dependente

In [ ]:
# Variável independente
X = df[['lstat']]

# Variável dependete
y = df['medv']

# Realizando a separação entre os dados que serão utilizados para treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)